File: backend/internal/content/series_tree.go
// Package content — series tree helpers for pamphlet organization.
//
// Pamphlets (EPAMs) group as series → chapters → pamphlet without a new
// Dynamo table. BuildSeriesTree walks ListByUser metadata and nests rows.
package content

import (
	"sort"
	"strings"
)

const (
	// UnassignedSeriesLabel is the bucket for pamphlets with empty series meta.
	UnassignedSeriesLabel = "(sin serie)"
	// UnassignedChapterLabel is the bucket for empty seriesChapter meta.
	UnassignedChapterLabel = "(sin capítulo)"
)

// SeriesTreeItem is one pamphlet leaf in the series tree.
type SeriesTreeItem struct {
	EpamID        string `json:"epamId"`
	Title         string `json:"title"`
	FileName      string `json:"fileName,omitempty"`
	Series        string `json:"series,omitempty"`
	SeriesChapter string `json:"seriesChapter,omitempty"`
	UpdatedAt     string `json:"updatedAt,omitempty"`
}

// SeriesTreeChapter holds pamphlets under one chapter name.
type SeriesTreeChapter struct {
	Name  string           `json:"name"`
	Items []SeriesTreeItem `json:"items"`
}

// SeriesTreeNode is one series with nested chapters.
type SeriesTreeNode struct {
	Name     string              `json:"name"`
	Chapters []SeriesTreeChapter `json:"chapters"`
}

// SeriesTreeResponse is the GET /api/epams/series-tree payload.
type SeriesTreeResponse struct {
	Count  int              `json:"count"`
	Series []SeriesTreeNode `json:"series"`
}

func seriesKey(raw string) string {
	s := strings.TrimSpace(raw)
	if s == "" {
		return UnassignedSeriesLabel
	}
	return s
}

func chapterKey(raw string) string {
	s := strings.TrimSpace(raw)
	if s == "" {
		return UnassignedChapterLabel
	}
	return s
}

// BuildSeriesTree groups epam metadata into series → chapters → items.
// Sorting is stable: series/chapter names ascending; items by title then epamId.
func BuildSeriesTree(records []EpamRecord) SeriesTreeResponse {
	type chapterBucket struct {
		items []SeriesTreeItem
	}
	seriesMap := map[string]map[string]*chapterBucket{}

	for _, rec := range records {
		sk := seriesKey(rec.Series)
		ck := chapterKey(rec.SeriesChapter)
		if seriesMap[sk] == nil {
			seriesMap[sk] = map[string]*chapterBucket{}
		}
		if seriesMap[sk][ck] == nil {
			seriesMap[sk][ck] = &chapterBucket{}
		}
		title := strings.TrimSpace(rec.Title)
		if title == "" {
			title = strings.TrimSpace(rec.FileName)
		}
		if title == "" {
			title = rec.EpamID
		}
		seriesMap[sk][ck].items = append(seriesMap[sk][ck].items, SeriesTreeItem{
			EpamID:        rec.EpamID,
			Title:         title,
			FileName:      rec.FileName,
			Series:        rec.Series,
			SeriesChapter: rec.SeriesChapter,
			UpdatedAt:     rec.UpdatedAt,
		})
	}

	seriesNames := make([]string, 0, len(seriesMap))
	for name := range seriesMap {
		seriesNames = append(seriesNames, name)
	}
	sort.Strings(seriesNames)

	out := make([]SeriesTreeNode, 0, len(seriesNames))
	total := 0
	for _, sName := range seriesNames {
		chaptersMap := seriesMap[sName]
		chapterNames := make([]string, 0, len(chaptersMap))
		for name := range chaptersMap {
			chapterNames = append(chapterNames, name)
		}
		sort.Strings(chapterNames)

		chapters := make([]SeriesTreeChapter, 0, len(chapterNames))
		for _, cName := range chapterNames {
			items := chaptersMap[cName].items
			sort.SliceStable(items, func(i, j int) bool {
				if items[i].Title != items[j].Title {
					return items[i].Title < items[j].Title
				}
				return items[i].EpamID < items[j].EpamID
			})
			total += len(items)
			chapters = append(chapters, SeriesTreeChapter{Name: cName, Items: items})
		}
		out = append(out, SeriesTreeNode{Name: sName, Chapters: chapters})
	}

	return SeriesTreeResponse{Count: total, Series: out}
}

File: frontend/src/lib/seriesTree.ts
/**
 * Series → chapter → pamphlet grouping.
 * Mirrors backend/internal/content/series_tree.go so /articulos and the
 * cloud open list share the same buckets and sort order.
 */

export const UNASSIGNED_SERIES_LABEL = "(sin serie)";
export const UNASSIGNED_CHAPTER_LABEL = "(sin capítulo)";

export type SeriesLeaf = {
  epamId: string;
  title: string;
  fileName?: string;
  series?: string;
  seriesChapter?: string;
  author?: string;
  date?: string;
  updatedAt?: string;
};

export type SeriesChapterNode = {
  name: string;
  items: SeriesLeaf[];
};

export type SeriesNode = {
  name: string;
  chapters: SeriesChapterNode[];
};

export type SeriesTree = {
  count: number;
  series: SeriesNode[];
};

export type SeriesSource = {
  epamId: string;
  title?: string;
  fileName?: string;
  series?: string;
  seriesChapter?: string;
  author?: string;
  date?: string;
  updatedAt?: string;
};

function seriesKey(raw: string | undefined): string {
  const s = (raw ?? "").trim();
  return s === "" ? UNASSIGNED_SERIES_LABEL : s;
}

function chapterKey(raw: string | undefined): string {
  const s = (raw ?? "").trim();
  return s === "" ? UNASSIGNED_CHAPTER_LABEL : s;
}

function leafTitle(rec: SeriesSource): string {
  const title = (rec.title ?? "").trim();
  if (title) return title;
  const fileName = (rec.fileName ?? "").trim();
  if (fileName) return fileName;
  return rec.epamId;
}

/** Group pamphlet metadata into series → chapters → items (stable sort). */
export function groupEpamsBySeries(records: SeriesSource[]): SeriesTree {
  const seriesMap = new Map<string, Map<string, SeriesLeaf[]>>();

  for (const rec of records) {
    const sk = seriesKey(rec.series);
    const ck = chapterKey(rec.seriesChapter);
    if (!seriesMap.has(sk)) seriesMap.set(sk, new Map());
    const chapters = seriesMap.get(sk)!;
    if (!chapters.has(ck)) chapters.set(ck, []);
    chapters.get(ck)!.push({
      epamId: rec.epamId,
      title: leafTitle(rec),
      fileName: rec.fileName,
      series: rec.series,
      seriesChapter: rec.seriesChapter,
      author: rec.author,
      date: rec.date,
      updatedAt: rec.updatedAt,
    });
  }

  const seriesNames = [...seriesMap.keys()].sort();
  const series: SeriesNode[] = [];
  let count = 0;
  for (const sName of seriesNames) {
    const chaptersMap = seriesMap.get(sName)!;
    const chapterNames = [...chaptersMap.keys()].sort();
    const chapters: SeriesChapterNode[] = [];
    for (const cName of chapterNames) {
      const items = chaptersMap.get(cName)!;
      items.sort((a, b) => {
        if (a.title !== b.title) return a.title < b.title ? -1 : 1;
        return a.epamId < b.epamId ? -1 : 1;
      });
      count += items.length;
      chapters.push({ name: cName, items });
    }
    series.push({ name: sName, chapters });
  }
  return { count, series };
}
